### ISSUE TITLE
this.base is replaced while component is mounted

### ISSUE THREAD

**Issue by kruczy** [developit/preact#243]
When a high order component is mounted its this.base gets replaced with a different element without thet component life-cycle events being called

There is no jsfiddle repro of this yet (and I tried) tho the problem occurs in my application consistently

What I know:
 - started with `5.1.0-beta.19` and still there in `5.5.0`
 - wrapping the component content in a div fixes the problem
 - the element to be rendered is a form element
 - wrapping the whole element in a form instead of a div breaks it
 - when render is called the first time `this.base` is a correct element (same as when mounting)
 - on second render `this.base` is the new element, it must be replaced after first render
 - the component is not unmounted at all
 - happens on full page refresh as well as router navigation

**Comment by kruczy** [comment]
@developit finally managed to reproduce it the same way it is happening in my app:
http://jsfiddle.net/mxrjewoh/1/

the console logged element should be the parent of the component although it is not

the reason for this will be that the element is of a different type div/span, is this expected?
If it is it would be awesome to be able to know from component perspective that this is happening to clenup/initialize the new element

**Comment by developit** [comment]
@kruczy - had typed up a reply elsewhere but apparently never sent it. Preact is exhibiting the same behavior as React here, so it is expected as you suspected - changing the type of the root element doesn't trigger a remount. If you are sure the base element is a DOM node, I'd suggest using refs since they are nulled when their corresponding elements are removed, which sounds pretty close to the DOM-centric lifecycle timing you are seeking. I know this isn't as useful if you cant count on that root node being a ref, though. For that, I have seen people pass the ref down through the tree as `ref_` where it can be applied to the innermost component's root node. 

A quick fix otherwise, would be to track `this.base` and react when it changes:

```js
componentDidUpdate() {
  if (this.base!=this.lastBase) {
    this.lastBase.removeEventListener(...);
    this.base.addEventListener(...);
    this.lastBase = this.base;
  }
}
```

As I looked into this I found a quirk with ref timing, specifically for the element name change at the component root case you noted here, so I'm looking to have a fix out for that shortly.
